dipy/tracking/local/localtrack.pyx

```diff
@@ -208,3 +210,254 @@ cdef int _local_tracker(
         # maximum length of streamline has been reached, return everything
         i = streamline.shape[0]
     return i
+
+
+def pft_tracker(
+        DirectionGetter dg,
+        ConstrainedTissueClassifier tc,
+        np.float_t[:] seed,
+        np.float_t[:] first_step,
+        np.float_t[:] voxel_size,
+        np.float_t[:,:] streamline,
+        np.float_t[:,:] directions,
+        double step_size,
+        int pft_nbr_back_steps,
+        int pft_max_steps,
+        int pft_max_trial,
+        int particle_count,
+        np.float_t[:,:,:,:] particle_paths,
+        np.float_t[:,:,:,:] particle_dirs,
+        np.float_t[:,:] particle_weights,
+        np.int_t[:,:,:]  particle_states):
+    cdef:
+        int i
+        TissueClass tissue_class
+
+    if (seed.shape[0] != 3 or first_step.shape[0] != 3 or
+            voxel_size.shape[0] != 3 or streamline.shape[1] != 3):
+        raise ValueError()
```

Comment:
IMO, this should be in the python code, probably in local_tracking, or simply removed (replaced by documentation?) One time would be ok but these checks are uselessly done thousands of times.

Reply:
This is done once per streamline. Each of these parameters is not specific to the tracker, but to the streamline that will be track. Moving it to the python code would be the same number of calls. Wouldn't 1,000,000 calls (streamlines) to this be negligible? What do you think?

Reply:
We don't know how well the branch prediction is working but we can probably assume it's working as well as it would on a C program, so it's 4 conditions for each streamlines. One of them is checking a constant (`voxel_size`) Another is checking is checking a 2D array (`streamline`) with a constant width. The 2 others are a 3D point (`seed`) and a 3D vector (`first_step`), which seems obvious.

My opinion on this is:
- Cython is horrible but fast, lets not do anything useless
- Duck Typing can mean different things, but there's a concept I often see in Python "The most Pythonic way to check the type of an object is... not to check it." 

But, hey, it's up to you! As you wrote, it's negligible compared to the rest of the processing :)